### PR TITLE
feat/timestamp

### PR DESCRIPTION
**Pull request by jenr24G**
adds timestamp checking to the canonical OCV query so as to avoid votes being counted that were cast before the start of the voting period

### CODE REVIEW

**Review by jstavh — APPROVED** [review]

**Comment by jenr24G** [comment]
"could not compile due to needless return" bite me clippy

**Comment by mxnkarou** [comment]
Tests need to be added as well.

**Comment by jenr24G** [comment]
kcov is broken. seems to be a problem with the CI

**Comment by jenr24G** [comment]
> Tests need to be added as well to maintain coverage.

This bit is already covered by tests. the code shouldn't add any new uncovered area, but i'll add whatever you think is necessary!

**Comment by jenr24G** [comment]
how tf does removing a space character make my run fail !?

File: .github/workflows/server.yml [hunk 1]
```diff
@@ -103,6 +103,7 @@ jobs:
 
       - name: Install KCov Dependencies
         run: |
+          sudo apt-get update && sudo apt dist-upgrade
```
Comment (mxnkarou) [inline .github/workflows/server.yml:106]: Is `dist-upgrade` necessary?

File: server/src/queries/mod.rs [hunk 2]
```diff
@@ -22,22 +22,50 @@ pub async fn get_latest_blockheight(
...
+            WHERE uc.type = 'payment'
+            AND uc.source_id = uc.receiver_id
+            AND uc.token = 1
+            AND NOT b.chain_status = 'orphaned'
+            AND buc.status = 'applied'
+            AND b.timestamp > (select extract(epoch from now())) - $1
+            "#, (2629800 * 3) as _
+            ).fetch_all(db).await
+        }
+    }
```
Comment (mxnkarou) [inline server/src/queries/mod.rs:70]: Can be shortened to:
```
    sqlx::query_as!(DBResponse,
        r#"
        SELECT DISTINCT pk.value as account, uc.memo as memo, uc.nonce as nonce, b.height as height, b.chain_status as "status: BlockStatus", b.timestamp as timestamp
        FROM user_commands AS uc
        JOIN blocks_user_commands AS buc
        ON uc.id = buc.user_command_id
        JOIN blocks AS b
        ON buc.block_id = b.id
        JOIN public_keys AS pk
        ON uc.source_id = pk.id
        WHERE uc.type = 'payment'
        AND uc.source_id = uc.receiver_id
        AND uc.token = 1
        AND NOT b.chain_status = 'orphaned'
        AND buc.status = 'applied'
        AND b.timestamp > $1
        "#, match timestamp {
            Some(ts) => ts,
            None => 2629800 * 3,
        }
        ).fetch_all(db).await

**Comment by mxnkarou** [comment]
Also - this implementation does not work - `http://localhost:8080/api/v1/test?network=Mainnet` returns 
`Failed to deserialize query string: invalid digit found in string`.

File: server/src/queries/mod.rs (hunk 2 above)
Comment (jenr24G) [reply server/src/queries/mod.rs:70]: no it cannot. if None is passed in the above code sample, the query will get all blocks until 3 months after the genesis block. we want 3 months before when the query is run, and I figured we would want to calculate "now" on the DB side, hence 2 different queries.

Comment (mxnkarou) [reply server/src/queries/mod.rs:70]: No not DB side - calculate it on the server & make that the default. (Still can be shortened then)

Comment (jenr24G) [reply server/src/queries/mod.rs:70]: > Also - this implementation is wrong - `http://localhost:8080/api/v1/test?network=Mainnet` returns `Failed to deserialize query string: invalid digit found in string`.
> 
> I recommend reading the Axum docs for query params handling.

ON the monday.com task, you asked for the timestamp to be passed as query parameter, then later you said in slack this query parameter should be a unix timestamp. The URL in this example does not have a `startDate` query parameter passed in

**Comment by jenr24G** [comment]
> Also - this implementation is wrong - `http://localhost:8080/api/v1/test?network=Mainnet` returns `Failed to deserialize query string: invalid digit found in string`.
> 
> I recommend reading the Axum docs for query params handling.

ON the monday.com task, you asked for the timestamp to be passed as query parameter, then later you said in slack this query parameter should be a unix timestamp. The URL in this example does not have a `startDate` query parameter passed in

File: server/src/queries/mod.rs (hunk 2 above)
Comment (mxnkarou) [reply server/src/queries/mod.rs:70]: i.E. something like:
```
        "#, match timestamp {
            Some(ts) => ts,
            None => std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs() as i64
            - 7889400,
}

**Comment by jenr24G** [comment]
got got by the ledgers again smh

File: server/src/queries/mod.rs (hunk 2 above)
Comment (jenr24G) [reply server/src/queries/mod.rs]: hey mine's a bit different but it does the same thing!

**Comment by jenr24G** [comment]
tested this latest version with Insomnia! No more deserialization problems!

**Comment by mxnkarou** [comment]
Passing in a timestamp works - default of 3 months doesn't work though.
No timestamp param seems to yield all signals as before.

**Comment by jenr24G** [comment]
I assumed the timestamps on the server were stored as seconds, they are actually milliseconds. We might want to change the deserialization of the timestamp field and query param to use `u128`s, like `Duration::as_millis()` does in order to avoid inevitable integer overflow

**Review by mxnkarou — APPROVED** [review]

**Pull request merged**